modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java

```diff
@@ -822,6 +839,73 @@ public boolean rebuildFromHashInProgress() {
         return commitUserIndex(ses, idxName);
     }
 
+    /**
+     * Checks and replace if provided column id matches value alias column.
+     *
+     * @param col Input column.
+     * @return Transformed result if needed.
+     */
+    private String aliasNameTransformer(IndexColumn col) {
+        if (desc.isKeyAliasColumn(col.column.getColumnId()))
+            return QueryUtils.KEY_FIELD_NAME;
+        else if (desc.isValueAliasColumn(col.column.getColumnId()))
+            return QueryUtils.VAL_FIELD_NAME;
+
+        return col.columnName;
+    }
+
+    /**
+     * Check index presence, throw exception if index with same name already exist or return {@code True} if
+     * index with same fields and search direction found.
+     *
+     * @param curIdx Index to check.
+     * @return {@code True} if equal index exist.
+     * @throws IgniteCheckedException If failed.
+     */
+    private @Nullable Index checkIndexPresence(Index curIdx) throws IgniteCheckedException {
+        IndexColumn[] curColumns = curIdx.getIndexColumns();
+
+        Index registredIdx = null;
+
+        for (Index idx : idxs) {
+            if (!(idx instanceof H2TreeIndex))
+                continue;
+
+            if (F.eq(curIdx.getName(), idx.getName()))
+                throw new IgniteCheckedException("Index already exists: " + idx.getName());
+
+            IndexColumn[] idxColumns = idx.getIndexColumns();
+
+            if (idxColumns.length == curColumns.length && idxColumns.length > 0) {
+                if (!F.eq(aliasNameTransformer(idxColumns[0]), aliasNameTransformer(curColumns[0])))
+                    continue;
+
+                boolean copy = true;
+
+                for (IndexColumn idxCol : idxColumns) {
+                    registredIdx = null;
+
+                    for (IndexColumn curCol : curColumns) {
+                        if (F.eq(aliasNameTransformer(idxCol), aliasNameTransformer(curCol))
```

Comment:
seems algorithm of checking indexes is not correct. Need to check columns by their position at index.